be/src/olap/schema_change.cpp

```diff
@@ -576,6 +583,248 @@ Status VSchemaChangeWithSorting::_external_sorting(vector<RowsetSharedPtr>& src_
+                                                   Field* field, const uint8_t* null_map,
+                                                   const uint8_t** ptr, size_t num_rows) {
+    size_t offset = 0;
+    auto next_run_step = [&]() {
+        size_t step = 1;
+        for (auto i = offset + 1; i < num_rows; ++i) {
+            if (null_map[offset] == null_map[i])
+                step++;
+            else
+                break;
+        }
+        return step;
+    };
+
+    try {
+        do {
+            auto step = next_run_step();
+            if (null_map[offset]) {
+                RETURN_IF_ERROR(_inverted_index_builders[index_writer_sign]->add_nulls(step));
+                *ptr += field->size() * step;
+            } else {
+                if (field->type() == FieldType::OLAP_FIELD_TYPE_ARRAY) {
+                    DCHECK(field->get_sub_field_count() == 1);
+                    const auto* col_cursor = reinterpret_cast<const CollectionValue*>(*ptr);
+                    RETURN_IF_ERROR(_inverted_index_builders[index_writer_sign]->add_array_values(
+                            field->get_sub_field(0)->size(), col_cursor, step));
+                } else {
+                    RETURN_IF_ERROR(_inverted_index_builders[index_writer_sign]->add_values(
+                            column_name, *ptr, step));
+                }
+            }
+            offset += step;
+        } while (offset < num_rows);
+    } catch (const std::exception& e) {
+        LOG(WARNING) << "CLuceneError occured: " << e.what();
+        return Status::Error<IO_ERROR>();
+    }
+
+    return Status::OK();
+}
+
+Status SchemaChangeForInvertedIndex::_add_data(const std::string& column_name,
+                                               const std::string& index_writer_sign, Field* field,
+                                               const uint8_t** ptr, size_t num_rows) {
+    try {
+        if (field->type() == FieldType::OLAP_FIELD_TYPE_ARRAY) {
+            DCHECK(field->get_sub_field_count() == 1);
+            const auto* col_cursor = reinterpret_cast<const CollectionValue*>(*ptr);
+            RETURN_IF_ERROR(_inverted_index_builders[index_writer_sign]->add_array_values(
+                    field->get_sub_field(0)->size(), col_cursor, num_rows));
+        } else {
+            RETURN_IF_ERROR(_inverted_index_builders[index_writer_sign]->add_values(
+                    column_name, *ptr, num_rows));
+        }
+    } catch (const std::exception& e) {
+        LOG(WARNING) << "CLuceneError occured: " << e.what();
+        return Status::Error<IO_ERROR>();
+    }
+
+    return Status::OK();
+}
+
+Status SchemaChangeForInvertedIndex::_write_inverted_index(int32_t segment_idx,
+                                                           vectorized::Block* block) {
+    LOG(INFO) << "begin to write inverted index";
+
+    // converter block data
+    vectorized::OlapBlockDataConvertor olap_data_convertor;
```

Comment:
It's not efficient to create an olap_data_convertor for each block.

Reply:
yes, you are right, I will fix it.